suite/tests/client-interface/drx-scattergather-instrumentation-fault.dll.c

```diff
@@ -0,0 +1,279 @@
+                    instr_is_gather(emulated_instr.instr);
+                if (data->is_scatter_gather) {
+                    data->app_addr = instr_get_app_pc(emulated_instr.instr);
+                    break;
+                }
+            }
+        }
+    }
+
+    *user_data = (void *)data;
+    return DR_EMIT_DEFAULT;
+}
+
+static dr_emit_flags_t
+event_app_instruction(void *drcontext, void *tag, instrlist_t *bb, instr_t *where,
+                      bool for_trace, bool translating, void *user_data)
+{
+    int i;
+    bool seen_memref = false;
+    instru_data_t *data = (instru_data_t *)user_data;
+
+    /* If the previous instruction was a scatter/gather expansion load/store,
+     * we insert another faulting store after it.
+     */
+    if (data->do_post_instruction_instrumentation) {
+        handle_post_instruction(drcontext, bb, where, data->app_addr);
+    }
+    data->do_post_instruction_instrumentation = false;
+
+    if (data->is_scatter_gather) {
+        instr_t *instr_operands = drmgr_orig_app_instr_for_operands(drcontext);
+        if (instr_operands != NULL &&
+            (instr_writes_memory(instr_operands) || instr_reads_memory(instr_operands))) {
+            DR_ASSERT(instr_is_app(instr_operands));
+            for (i = 0; i < instr_num_dsts(instr_operands); ++i) {
+                const opnd_t dst = instr_get_dst(instr_operands, i);
+                if (opnd_is_memory_reference(dst)) {
+                    if (seen_memref) {
+                        DR_ASSERT_MSG(false,
+                                      "Found inst with multiple memory destinations");
+                        break;
+                    }
+                    insert_faulting_store(drcontext, bb, where, data->app_addr);
+                    seen_memref = true;
+                    data->do_post_instruction_instrumentation = true;
+                }
+            }
+            for (i = 0; i < instr_num_srcs(instr_operands); ++i) {
+                const opnd_t src = instr_get_src(instr_operands, i);
+                if (opnd_is_memory_reference(src)) {
+                    if (seen_memref) {
+                        DR_ASSERT_MSG(false,
+                                      "Found inst with multiple memory destinations");
+                        break;
+                    }
+                    insert_faulting_store(drcontext, bb, where, data->app_addr);
+                    seen_memref = true;
+                    data->do_post_instruction_instrumentation = true;
+                }
+            }
+        }
+    }
+
+    if (drmgr_is_last_instr(drcontext, where))
+        dr_thread_free(drcontext, data, sizeof(*data));
+    return DR_EMIT_DEFAULT;
+}
+
+static dr_emit_flags_t
+event_bb_app2app(void *drcontext, void *tag, instrlist_t *bb, bool for_trace,
+                 bool translating)
+{
+    if (!drx_expand_scatter_gather(drcontext, bb, NULL)) {
+        DR_ASSERT(false);
+    }
+    drx_tail_pad_block(drcontext, bb);
```

Comment:
Why is this needed?

Reply:
I don't know. This client started out as a stripped down memval_simple and memval_simple has it. I have removed it.

Reply:
IIRC this is required for the post-store-instruction instrumentation used by memval and without it it will fail (probably assert in debug build; not sure about release), if this test has the same post instrumentation.

Reply:
We removed the post-instrumentation as that did not add any additional coverage to the test.